src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs

```diff
@@ -0,0 +1,353 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions.LanguageServices
+{
+    /// <summary>
+    /// Helper class to detect regex pattern tokens in a document efficiently.
+    /// </summary>
+    internal sealed class RegexPatternDetector
+    {
+        private const string _patternName = "pattern";
+
+        private static readonly ConditionalWeakTable<SemanticModel, RegexPatternDetector> _modelToDetector =
+            new ConditionalWeakTable<SemanticModel, RegexPatternDetector>();
+
+        private readonly RegexEmbeddedLanguage _language;
+        private readonly SemanticModel _semanticModel;
+        private readonly INamedTypeSymbol _regexType;
+        private readonly HashSet<string> _methodNamesOfInterest;
+
+        /// <summary>
+        /// Helps match patterns of the form: language=regex,option1,option2,option3
```

Comment:
>language=regex,option1,option2,option3 [](start = 46, length = 38)

It's probably good to cover this more. I didn't see tests with options. When we do, let's include spaces/tabs/etc.

Reply:
Sure

Reply:
(done)